File: frontend/src/services/schemas.ts
import { z } from 'zod'

/**
 * Zod şemaları — FastAPI (app/api/schemas.py) ile birebir uyumlu.
 * Backend tarafında değişiklik olursa burayı da güncelle.
 */

export const CoordinateSchema = z.object({
  lat: z.number().min(-90).max(90),
  lon: z.number().min(-180).max(180),
})
export type Coordinate = z.infer<typeof CoordinateSchema>

export const HealthResponseSchema = z.object({
  status: z.literal('ok'),
  service: z.string(),
  version: z.string(),
  model_available: z.boolean(),
  model_version: z.string().nullable().optional(),
  vehicle_count: z.number(),
})
export type HealthResponse = z.infer<typeof HealthResponseSchema>

export const VehicleSummarySchema = z.object({
  id: z.string(),
  name: z.string(),
  make: z.string(),
  model: z.string(),
  variant: z.string(),
  year: z.number(),
  body_type: z.string(),
  usable_battery_kwh: z.number(),
  ideal_consumption_wh_km: z.number(),
  wltp_range_km: z.number(),
  max_dc_charge_kw: z.number(),
})
export type VehicleSummary = z.infer<typeof VehicleSummarySchema>

export const VehicleDetailSchema = VehicleSummarySchema.extend({
  drivetrain: z.string(),
  battery_chemistry: z.string(),
  gross_battery_kwh: z.number(),
  soc_min_pct: z.number(),
  soc_max_pct: z.number(),
  regen_efficiency: z.number(),
  weight_kg: z.number(),
  max_ac_charge_kw: z.number(),
  temp_penalty_factor: z.number(),
  charge_curve_hint: z.string(),
  default_hvac_load_kw: z.number(),
})
export type VehicleDetail = z.infer<typeof VehicleDetailSchema>

export const StrategyNameSchema = z.enum(['fast', 'efficient', 'balanced'])
export type StrategyName = z.infer<typeof StrategyNameSchema>

export const RecommendedStopSchema = z.object({
  name: z.string(),
  distance_along_route_km: z.number(),
  detour_distance_km: z.number().default(0),
  detour_minutes: z.number().default(0),
  arrival_soc_percent: z.number().default(0),
  target_soc_percent: z.number().default(0),
  charge_minutes: z.number().default(0),
  power_kw: z.number().default(0),
})
export type RecommendedStop = z.infer<typeof RecommendedStopSchema>

export const ProfileCardSchema = z.object({
  key: StrategyNameSchema,
  label: z.string(),
  feasible: z.boolean(),
  total_energy_kwh: z.number().nullable().optional(),
  total_trip_minutes: z.number().nullable().optional(),
  charging_minutes: z.number().nullable().optional(),
  stop_count: z.number().nullable().optional(),
  final_soc_pct: z.number().nullable().optional(),
  used_ml: z.boolean(),
  model_version: z.string().nullable().optional(),
  recommended_stops: z.array(RecommendedStopSchema).default([]),
  raw: z.record(z.string(), z.unknown()).default({}),
})
export type ProfileCard = z.infer<typeof ProfileCardSchema>

export const OptimizeResponseSchema = z.object({
  status: z.string(),
  vehicle_id: z.string(),
  vehicle_name: z.string(),
  initial_soc_pct: z.number(),
  final_soc_pct: z.number(),
  total_distance_km: z.number(),
  total_energy_kwh: z.number(),
  used_ml: z.boolean(),
  ml_segment_count: z.number().default(0),
  heuristic_segment_count: z.number().default(0),
  model_version: z.string().nullable().optional(),
  recommended_profile: StrategyNameSchema.nullable().optional(),
  profiles: z.array(ProfileCardSchema),
  raw_optimization: z.record(z.string(), z.unknown()).default({}),
})
export type OptimizeResponse = z.infer<typeof OptimizeResponseSchema>

export const OptimizeRequestSchema = z.object({
  vehicle_id: z.string().min(1),
  start: CoordinateSchema,
  end: CoordinateSchema,
  initial_soc_pct: z.number().min(0).max(100),
  target_arrival_soc_pct: z.number().min(0).max(100).nullable().optional(),
  strategies: z.array(StrategyNameSchema).min(1),
  use_ml: z.boolean().default(false),
})
export type OptimizeRequest = z.infer<typeof OptimizeRequestSchema>

export const SpeedLimitSegmentSchema = z.object({
  start_index: z.number(),
  end_index: z.number(),
  maxspeed_kmh: z.number().nullable().optional(),
  highway: z.string().nullable().optional(),
})
export type SpeedLimitSegment = z.infer<typeof SpeedLimitSegmentSchema>

export const SpeedLimitsResponseSchema = z.object({
  segments: z.array(SpeedLimitSegmentSchema),
  source: z.string(),
  sampled_point_count: z.number().default(0),
})
export type SpeedLimitsResponse = z.infer<typeof SpeedLimitsResponseSchema>

export const SpeedLimitsRequestSchema = z.object({
  geometry: z.array(z.array(z.number())).min(2),
  sample_every_n_points: z.number().int().positive().default(20),
})
export type SpeedLimitsRequest = z.infer<typeof SpeedLimitsRequestSchema>

export const GeocodeResultItemSchema = z.object({
  display_name: z.string(),
  name: z.string(),
  lat: z.number(),
  lon: z.number(),
  type: z.string().nullable().optional(),
  importance: z.number().default(0),
  country_code: z.string().nullable().optional(),
})
export type GeocodeResultItem = z.infer<typeof GeocodeResultItemSchema>

export const GeocodeResponseSchema = z.object({
  query: z.string(),
  results: z.array(GeocodeResultItemSchema),
})
export type GeocodeResponse = z.infer<typeof GeocodeResponseSchema>

export const RouteResponseSchema = z.object({
  summary: z.object({
    distance_km: z.number(),
    duration_min: z.number(),
    geometry_point_count: z.number(),
    sampled_point_count: z.number(),
    weather_point_count: z.number(),
    station_count: z.number(),
    avg_temp_c: z.number().nullable().optional(),
    min_temp_c: z.number().nullable().optional(),
    max_temp_c: z.number().nullable().optional(),
    avg_grade_pct: z.number().nullable().optional(),
    max_uphill_grade_pct: z.number().nullable().optional(),
    max_downhill_grade_pct: z.number().nullable().optional(),
  }),
  geometry: z.array(z.array(z.number())),
  elevation_profile: z.array(z.record(z.string(), z.unknown())).default([]),
  slope_segments: z.array(z.record(z.string(), z.unknown())).default([]),
  weather: z.record(z.string(), z.unknown()).default({}),
  stations: z.array(z.record(z.string(), z.unknown())).default([]),
})
export type RouteResponse = z.infer<typeof RouteResponseSchema>

File: app/api/schemas.py
"""
Pydantic v2 request/response şemaları.

Tek dosyada tüm API modelleri; endpoint'ler arası paylaşımı kolaylaştırır.
"""

from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator


# =========================================================
# Ortak
# =========================================================


class Coordinate(BaseModel):
    """Enlem/boylam çifti."""

    model_config = ConfigDict(extra="forbid")

    lat: float = Field(..., ge=-90.0, le=90.0, description="Enlem (-90, 90)")
    lon: float = Field(..., ge=-180.0, le=180.0, description="Boylam (-180, 180)")

    def as_tuple(self) -> tuple[float, float]:
        return (self.lat, self.lon)


class ErrorResponse(BaseModel):
    error: str
    detail: Optional[str] = None


class HealthResponse(BaseModel):
    status: Literal["ok"] = "ok"
    service: str = "ev-route-optimizer"
    version: str
    model_available: bool
    model_version: Optional[str] = None
    vehicle_count: int


# =========================================================
# Araçlar
# =========================================================


class VehicleSummary(BaseModel):
    id: str
    name: str
    make: str
    model: str
    variant: str
    year: int
    body_type: str
    usable_battery_kwh: float
    ideal_consumption_wh_km: float
    wltp_range_km: float
    max_dc_charge_kw: float


class VehicleDetail(VehicleSummary):
    drivetrain: str
    battery_chemistry: str
    gross_battery_kwh: float
    soc_min_pct: float
    soc_max_pct: float
    regen_efficiency: float
    weight_kg: float
    max_ac_charge_kw: float
    temp_penalty_factor: float
    charge_curve_hint: str
    default_hvac_load_kw: float


# =========================================================
# /route
# =========================================================


class RouteRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    start: Coordinate
    end: Coordinate
    elevation_min_spacing_km: float = Field(5.0, gt=0)
    elevation_max_points: int = Field(60, gt=0, le=500)
    weather_sample_limit: int = Field(5, gt=0, le=50)
    station_query_every_n_points: int = Field(4, gt=0)
    station_distance_km: float = Field(5.0, gt=0)
    station_max_results_per_query: int = Field(10, gt=0, le=100)
    allow_station_fallback: bool = True


class RouteSummary(BaseModel):
    distance_km: float
    duration_min: float
    geometry_point_count: int
    sampled_point_count: int
    weather_point_count: int
    station_count: int
    avg_temp_c: Optional[float] = None
    min_temp_c: Optional[float] = None
    max_temp_c: Optional[float] = None
    avg_grade_pct: Optional[float] = None
    max_uphill_grade_pct: Optional[float] = None
    max_downhill_grade_pct: Optional[float] = None


class RouteResponse(BaseModel):
    summary: RouteSummary
    geometry: List[List[float]] = Field(
        default_factory=list,
        description="Rota koordinatları [[lat,lon], ...]",
    )
    elevation_profile: List[Dict[str, Any]] = Field(default_factory=list)
    slope_segments: List[Dict[str, Any]] = Field(default_factory=list)
    weather: Dict[str, Any] = Field(default_factory=dict)
    stations: List[Dict[str, Any]] = Field(default_factory=list)


# =========================================================
# /stations
# =========================================================


class StationConnection(BaseModel):
    connection_type: Optional[str] = None
    power_kw: Optional[float] = None
    current_type: Optional[str] = None
    quantity: Optional[int] = None
    is_fast_charge_capable: Optional[bool] = None
    status: Optional[str] = None


class StationSummary(BaseModel):
    ocm_id: int
    name: str
    operator: Optional[str] = None
    address: str
    town: Optional[str] = None
    latitude: float
    longitude: float
    distance_km: Optional[float] = None
    number_of_points: Optional[int] = None
    is_operational: Optional[bool] = None
    connections: List[StationConnection] = Field(default_factory=list)


# =========================================================
# /estimate-consumption
# =========================================================


class SegmentInput(BaseModel):
    model_config = ConfigDict(extra="forbid")

    segment_length_km: float = Field(..., gt=0)
    avg_speed_kmh: float = Field(..., gt=0, le=300)
    elevation_gain_m: float = Field(0.0, ge=0)
    elevation_loss_m: float = Field(0.0, ge=0)
    temperature_c: Optional[float] = Field(None, ge=-50, le=60)
    soc_start_percent: float = Field(80.0, ge=0, le=100)
    soc_end_percent: Optional[float] = Field(None, ge=0, le=100)


class WeatherInput(BaseModel):
    model_config = ConfigDict(extra="forbid")

    temperature_c: Optional[float] = None


class EstimateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    vehicle_id: str
    segment: SegmentInput
    weather: Optional[WeatherInput] = None


class EstimateResponse(BaseModel):
    vehicle_id: str
    source: str = Field(..., description="ml | heuristic_* | fallback")
    used_model: bool
    predicted_energy_kwh: float
    fallback_energy_kwh: float
    model_version: Optional[str] = None
    features: Dict[str, Any] = Field(default_factory=dict)
    error: Optional[str] = None


# =========================================================
# /optimize
# =========================================================


StrategyName = Literal["fast", "efficient", "balanced"]


class OptimizeRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    vehicle_id: str
    start: Coordinate
    end: Coordinate
    initial_soc_pct: float = Field(..., ge=0.0, le=100.0)
    strategies: List[StrategyName] = Field(
        default_factory=lambda: ["fast", "efficient", "balanced"]
    )
    use_ml: bool = False

    # /route parametreleri — opsiyonel override
    elevation_min_spacing_km: float = Field(5.0, gt=0)
    elevation_max_points: int = Field(60, gt=0, le=500)
    station_distance_km: float = Field(5.0, gt=0)

    @field_validator("strategies")
    @classmethod
    def _strategies_non_empty_unique(cls, v: List[str]) -> List[str]:
        if not v:
            raise ValueError("En az 1 strateji seçilmeli.")
        deduped: List[str] = []
        for strat in v:
            if strat not in deduped:
                deduped.append(strat)
        return deduped


class ProfileCard(BaseModel):
    key: StrategyName
    label: str
    feasible: bool
    total_energy_kwh: Optional[float] = None
    total_trip_minutes: Optional[float] = None
    charging_minutes: Optional[float] = None
    stop_count: Optional[int] = None
    final_soc_pct: Optional[float] = None
    used_ml: bool = False
    model_version: Optional[str] = None
    raw: Dict[str, Any] = Field(default_factory=dict)


class OptimizeResponse(BaseModel):
    status: str
    vehicle_id: str
    vehicle_name: str
    initial_soc_pct: float
    final_soc_pct: float
    total_distance_km: float
    total_energy_kwh: float
    used_ml: bool
    ml_segment_count: int = 0
    heuristic_segment_count: int = 0
    model_version: Optional[str] = None
    recommended_profile: Optional[StrategyName] = None
    profiles: List[ProfileCard] = Field(default_factory=list)
    raw_optimization: Dict[str, Any] = Field(default_factory=dict)


# =========================================================
# /speed-limits
# =========================================================


class SpeedLimitsRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    geometry: List[List[float]] = Field(
        ...,
        description="Rota koordinat listesi [[lat, lon], ...]",
        min_length=2,
    )
    sample_every_n_points: int = Field(
        20,
        gt=0,
        le=200,
        description="Hız sınırı sorgusu için her N noktada bir örnekle.",
    )


class SpeedLimitSegment(BaseModel):
    start_index: int
    end_index: int
    maxspeed_kmh: Optional[float] = None
    highway: Optional[str] = None


class SpeedLimitsResponse(BaseModel):
    segments: List[SpeedLimitSegment] = Field(default_factory=list)
    source: str = Field(default="overpass", description="overpass | fallback")
    sampled_point_count: int = 0
